File: dep/frontend/src/action/connect_nodes.ts
import { String } from "graphflo";
import { Action, actionStrId } from "guifast_shared";

export const connectNodesStr = "connect_nodes";
export const connectNodesId = actionStrId(String.module, connectNodesStr);

export interface ConnectNodes extends Action {
    readonly start_connector_index: number;
    readonly finish_connector_index: number;
}

export const connectNodes = (
    startConnectorIndex: number,
    finishConnectorIndex: number
): ConnectNodes => {
    return {
        type: connectNodesId,
        start_connector_index: startConnectorIndex,
        finish_connector_index: finishConnectorIndex
    };
};

File: dep/backend/src/func/dispatch/connect_nodes.rs
use { GRAPHFLO, guifast_shared };
use action::{ ConnectNodes, ConnectNodesResponse };
use error::*;

pub unsafe fn connect_nodes(action: &ConnectNodes) -> Result<()> {
    let graphflo = GRAPHFLO.read()?;

    graphflo
        .get_graph_mut()?
        .connect_nodes(action.start_connector_index, action.finish_connector_index)?;

    let connect_nodes_response = ConnectNodesResponse::new(action.start_connector_index, action.finish_connector_index);
    guifast_shared::send_to_shared_store(&connect_nodes_response)?;
    Ok(())
}
